README.md
[![Yii2](https://img.shields.io/badge/required-Yii2_v2.0.40-blue.svg)](https://packagist.org/packages/yiisoft/yii2)
[![Downloads](https://img.shields.io/packagist/dt/wdmg/yii2-api.svg)](https://packagist.org/packages/wdmg/yii2-api)
[![Packagist Version](https://img.shields.io/packagist/v/wdmg/yii2-api.svg)](https://packagist.org/packages/wdmg/yii2-api)
![Progress](https://img.shields.io/badge/progress-ready_to_use-green.svg)
[![GitHub license](https://img.shields.io/github/license/wdmg/yii2-api.svg)](https://github.com/wdmg/yii2-api/blob/master/LICENSE)

<img src="./docs/images/yii2-api.png" width="100%" alt="Yii2 API module" />

# Yii2 API
API control module for Yii2.
This module is an integral part of the [Butterfly.СMS](https://butterflycms.com/) content management system, but can also be used as an standalone extension.
Copyrights (c) 2019-2023 [W.D.M.Group, Ukraine](https://wdmg.com.ua/)

# Requirements 
* PHP 5.6 or higher
* Yii2 v.2.0.40 and newest
* [Yii2 Base](https://github.com/wdmg/yii2-base) module (required)
* [Yii2 Users](https://github.com/wdmg/yii2-users) module (required)
* [Yii2 Options](https://github.com/wdmg/yii2-options) module (support)
* [Yii2 SelectInput](https://github.com/wdmg/yii2-selectinput) widget
* [ClipboardJS](https://github.com/zenorocha/clipboard.js) asset library (required)

# Installation
To install the module, run the following command in the console:

`$ composer require "wdmg/yii2-api"`

After configure db connection, run the following command in the console:

`$ php yii api/init`

And select the operation you want to perform:
  1) Apply all module migrations
  2) Revert all module migrations

# Migrations
In any case, you can execute the migration and create the initial data, run the following command in the console:

`$ php yii migrate --migrationPath=@vendor/wdmg/yii2-api/migrations`

# Configure
To add a module to the project, add the following data in your configuration file:

    'modules' => [
        ...
        'api' => [
            'class' => 'wdmg\api\Module',
            'routePrefix' => 'admin', // routing prefix for dashboard
            'accessTokenExpire', => 3600 // lifetime of `access_token`, `0` - unlimited
            'blockedIp' => [], // array, blocked access from IP`s
            'rateLimit' => 30, // request`s to API per minute
            'rateLimitHeaders' => false, // send HTTP-headers of rate limit
            'sendAccessToken' => true, // send access token with HTTP-headers
            'authMethods' => [ // auth methods to allow
                'basicAuth' => true,
                'bearerAuth' => true,
                'paramAuth' => true
            ],
            'authMethods' => [ // allowed API modes
                'public' => true,
                'private' => true
            ],
            'allowedApiModels' => [ // allowed API models
                'public' => [
                    "wdmg\api\models\api\NewsAPI" => true,
                    "wdmg\api\models\api\OptionsAPI" => false,
                    "wdmg\api\models\api\PagesAPI" => false,
                    "wdmg\api\models\api\UsersAPI" => false,
                    ...
                ],
                'private' => [
                    "wdmg\api\models\api\NewsAPI" => true,
                    "wdmg\api\models\api\OptionsAPI" => true,
                    "wdmg\api\models\api\PagesAPI" => true,
                    "wdmg\api\models\api\UsersAPI" => true,
                    ...
                ],
            ]
        ],
        ...
    ],

# Usecase

Request to API with base auth by username and password (option `authMethods['basicAuth']` must be set to `true`):

    $ curl 'http://example.com/api/users' \
    -XGET \
    -H 'Accept: application/json, text/javascript, */*; q=0.01' \
    -H 'Authorization: Basic YOUR_USERNAME_AND_PASSWORD'
    
<b>Attention!</b> YOUR_USERNAME_AND_PASSWORD in format `username:password` and has been encoded, like `base64_encode('username:password')`
After successful authorization, the server will return `X-Access-Token` for use in other requests (option `sendAccessToken` must be set to `true`).


Request to API with query param `access_token` (option `authMethods['paramAuth']` must be set to `true`):

    $ curl 'http://example.com/api/users?access-token=YOUR_API_ACCESS_TOKEN' \
    -XGET \
    -H 'Accept: application/json, text/javascript, */*; q=0.01'

Request to API with bearer `access_token` (option `authMethods['bearerAuth']` must be set to `true`):

    $ curl 'http://example.com/api/users' \
    -XGET \
    -H 'Accept: application/json, text/javascript, */*; q=0.01' \
    -H 'Authorization: Bearer YOUR_API_ACCESS_TOKEN'

If the access token has expired, the server will return new `X-Access-Token` for use in other requests (option `sendAccessToken` must be set to `true`).

# Routing

Admin dashboard path by default: http://example.com/admin/api/

Path for access to API endpoint: http://example.com/api/<code>model/module/action</code>
Also see official guideline: https://github.com/yiisoft/yii2/blob/master/docs/guide/rest-quick-start.md

Use the `Module::dashboardNavItems()` method of the module to generate a navigation items list for NavBar, like this:

    <?php
        echo Nav::widget([
        'options' => ['class' => 'navbar-nav navbar-right'],
            'label' => 'Modules',
            'items' => [
                Yii::$app->getModule('api')->dashboardNavItems(),
                ...
            ]
        ]);
    ?>

# Status and version [ready to use]
* v.2.0.1 - Add log activity, fix users API save/update
* v.2.0.0 - Basic Auth fixed
* v.1.4.0 - Update dependencies, copyrights
* v.1.3.12 - Added base API for Menu module, request helper
* v.1.3.11 - Update README.md and dependencies
* v.1.3.10 - Added base API for Comments module

CHANGELOG.md
Changelog
=========

## 2.0.*
 * Fix nav menu for dashboard

## 2.0.1 (2023-06-24)
 * Add log activity
 * Fix users API save/update
 * Fixed permissions for Users API

## 2.0.0 (2023-06-18)
 * Basic Auth fixed

## 1.4.0 (2023-05-30)
 * Update dependencies, copyrights
 * MenuAPI fixed

## 1.3.12 (2020-12-24)
 * Request helper for testing area
 * Added base API for Menu module
 
## 1.3.11 (2020-06-14)
 * Update README.md and dependencies
 
## 1.3.10 (2020-06-08)
 * Added Highlight for response API
 * Added base API for Comments module
 * Disable strictParsing for UrlManager

## 1.3.9 (2020-04-28)
 * Update README.md and dependencies
 
## 1.3.8 (2020-04-15)
 * Extend fields
 * Added base API for Content module
 
## 1.3.7 (2020-03-25)
 * Added pagination
 * Up to date dependencies
 
## 1.3.6 (2020-03-21)
 * Added base API for Media module
 * Expand API for Blog module

## 1.3.5 (2020-03-07)
 * Added base API for Blog module
 
## 1.3.4 (2020-02-25)
 * Added base API for Search module
 
## 1.3.3 (2019-12-10)
 * Fixed deprecated class declaration

## 1.3.2 (2019-12-03)
 * Added base API for Newsletters and Subscribers modules
 
## 1.3.1 (2019-10-15)
 * Fix context and translations
 
## 1.3.0 (2019-10-10)
 * Separation of access to public and private API
 
## 1.2.13 (2019-08-06)
 * Added base API for News and mailer modules
 
## 1.2.12 (2019-07-30)
 * Added base API for Pages module
 
## 1.2.11 (2019-07-16)
 * Added extra options to composer.json and navbar menu icon

## 1.2.10 (2019-07-07)
 * Added base API for Redirects module
 
## 1.2.9 (2019-06-09)
 * Added choice param for non interactive mode
 
## 1.2.8 (2019-06-07)
 * Module refactoring
 
## 1.2.7 (2019-06-05)
 * Update Yii2 version

## 1.2.6 (2019-06-04)
 * Update dependencies
 * Module transferred to base module interface
 
## 1.2.4 (2019-06-04)
 * Resolved bug in submodules names

## 1.2.3 (2019-06-01)
 * Bugfix, routing and refactoring
 
## 1.2.2 (2019-05-31)
 * Update Yii2 version
 
## 1.2.1 (2019-05-31)
 * Send access token with HTTP-headers
 
## 1.2.0 (2019-05-30)
 * Added API-modules
 * Added views and CRUD
 
## 1.1.0 (2019-05-30)
 * Added auth methods
 
## 1.0.2 (2019-05-28)
 * Bugfix for console operations other modules.
 * Added module name and description, translations. Routing fix.
 
## 1.0.1 (2019-05-28)
 * Added check access by IP
 
## 1.0.0 (2019-05-27)
 * Added rate limit for API-requests
 * Added API-auth by access-token in custom IdentityInterface
 * Added separate controllers and models for API
 
## 0.0.3 (2019-05-27)
 * Added routing and base auth
 
## 0.0.2 (2019-05-26)
 * Added first (test) API for Module::Users
 
## 0.0.1 (2019-05-25)
 * Added composer package
 * Added bootstrap, base module
 * Added base migrations

Module.php
<?php

namespace wdmg\api;

/**
 * Yii2 API
 *
 * @category        Module
 * @version         2.0.1
 * @author          Alexsander Vyshnyvetskyy <alex.vyshnyvetskyy@gmail.com>
 * @link            https://github.com/wdmg/yii2-api
 * @copyright       Copyright (c) 2019 - 2023 W.D.M.Group, Ukraine
 * @license         https://opensource.org/licenses/MIT Massachusetts Institute of Technology (MIT) License
 *
 */

use wdmg\helpers\ArrayHelper;
use Yii;
use wdmg\base\BaseModule;

/**
 * api module definition class
 */
class Module extends BaseModule
{
    /**
     * {@inheritdoc}
     */
    public $controllerNamespace = 'wdmg\api\controllers';

    /**
     * {@inheritdoc}
     */
    public $defaultRoute = 'api/index';

    /**
     * @var string, the name of module
     */
    public $name = "API";

    /**
     * @var string, the description of module
     */
    public $description = "API control module with testing tool";

    /**
     * @var string the module version
     */
    private $version = "2.0.1";

    /**
     * @var integer, priority of initialization
     */
    private $priority = 10;

    /**
     * @var integer, lifetime of `acces_token` by default, `0` - unlimited
     */
    public $accessTokenExpire = 3600;

    /**
     * @var integer, request`s to API per minute by default
     */
    public $rateLimit = 30;

    /**
     * @var boolean, sent rate limit with HTTP-headers
     */
    public $rateLimitHeaders = false;

    /**
     * @var array of the allowed auth methods
     */
    public $authMethods = [
        'basicAuth' => true,
        'bearerAuth' => true,
        'paramAuth' => true
    ];

    /**
     * @var array of the allowed API modes
     */
    public $allowedApiModes = [
        'public' => true,
        'private' => true
    ];

    /**
     * @var array of the allowed API models
     */
    public $allowedApiModels = [
        'public' => [
            "wdmg\api\models\api\MailerAPI" => false,
            "wdmg\api\models\api\NewsAPI" => true,
            "wdmg\api\models\api\BlogAPI" => true,
            "wdmg\api\models\api\OptionsAPI" => false,
            "wdmg\api\models\api\ContentAPI" => true,
            "wdmg\api\models\api\PagesAPI" => true,
            "wdmg\api\models\api\MediaAPI" => true,
            "wdmg\api\models\api\SearchAPI" => true,
            "wdmg\api\models\api\MenuAPI" => true,
            "wdmg\api\models\api\CommentsAPI" => true,
            "wdmg\api\models\api\LiveSearchAPI" => false,
            "wdmg\api\models\api\RedirectsAPI" => false,
            "wdmg\api\models\api\StatsAPI" => false,
            "wdmg\api\models\api\TasksAPI" => false,
            "wdmg\api\models\api\TicketsAPI" => false,
            "wdmg\api\models\api\UsersAPI" => false,
            "wdmg\api\models\api\SubscribersAPI" => false,
            "wdmg\api\models\api\SubscribersListAPI" => false,
            "wdmg\api\models\api\NewslettersAPI" => false,
        ],
        'private' => [
            "wdmg\api\models\api\MailerAPI" => true,
            "wdmg\api\models\api\NewsAPI" => true,
            "wdmg\api\models\api\BlogAPI" => true,
            "wdmg\api\models\api\OptionsAPI" => true,
            "wdmg\api\models\api\ContentAPI" => true,
            "wdmg\api\models\api\PagesAPI" => true,
            "wdmg\api\models\api\MediaAPI" => true,
            "wdmg\api\models\api\SearchAPI" => true,
            "wdmg\api\models\api\MenuAPI" => true,
            "wdmg\api\models\api\CommentsAPI" => true,
            "wdmg\api\models\api\LiveSearchAPI" => true,
            "wdmg\api\models\api\RedirectsAPI" => true,
            "wdmg\api\models\api\StatsAPI" => true,
            "wdmg\api\models\api\TasksAPI" => true,
            "wdmg\api\models\api\TicketsAPI" => true,
            "wdmg\api\models\api\UsersAPI" => true,
            "wdmg\api\models\api\SubscribersAPI" => true,
            "wdmg\api\models\api\SubscribersListAPI" => true,
            "wdmg\api\models\api\NewslettersAPI" => true,
        ],
    ];

    /**
     * @var boolean, send access token with HTTP-headers
     */
    public $sendAccessToken = true;

    /**
     * @var array, blocked access from IP`s
     */
    public $blockedIp = [];

	/**
	 * @var array, the list of support locales for multi-language versions of posts.
	 * @note This variable will be override if you use the `wdmg\yii2-translations` module.
	 */
	public $supportLocales = ['ru-RU', 'uk-UA', 'en-US'];

    /**
     * {@inheritdoc}
     */
    public function init()
    {
        parent::init();

        // Set version of current module
        $this->setVersion($this->version);

        // Set priority of current module
        $this->setPriority($this->priority);
    }

    /**
     * {@inheritdoc}
     */
    public function dashboardNavItems($options = null)
    {
        $items = [
            'label' => $this->name,
            'url' => '#',
            'icon' => 'fa fa-fw fa-plug',
            'active' => in_array(\Yii::$app->controller->module->id, [$this->id]),
            'items' => [
                [
                    'label' => Yii::t('app/modules/api', 'List of API`s'),
                    'url' => [$this->routePrefix . '/api/'],
                    'active' => (in_array(\Yii::$app->controller->module->id, ['api']) &&  Yii::$app->controller->id == 'api'),
                ],
                [
                    'label' => Yii::t('app/modules/api', 'Access to API`s'),
                    'url' => [$this->routePrefix . '/api/access/'],
                    'active' => (in_array(\Yii::$app->controller->module->id, ['api']) &&  Yii::$app->controller->id == 'access'),
                ],
            ]
        ];

	    if (!is_null($options)) {

		    if (isset($options['count'])) {
			    $items['label'] .= '<span class="badge badge-default float-right">' . $options['count'] . '</span>';
			    unset($options['count']);
		    }

		    if (is_array($options))
			    $items = ArrayHelper::merge($items, $options);

	    }

	    return $items;
    }

    /**
     * {@inheritdoc}
     */
    public function bootstrap($app)
    {
        // Get the module instance
        //$module = $app->getModule('admin/api');
        $module = $this;

        // Configure urlManager and Request component
        if (!Yii::$app instanceof \yii\console\Application) {
            $app->getRequest()->parsers[] = ['application/json' => 'yii\web\JsonParser'];
            $app->getRequest()->enableCookieValidation = false;
            $app->getRequest()->enableCsrfValidation = false;
	        $app->getRequest()->enableCsrfCookie = false;
            Yii::$app->response->cookies->removeAll();

            // Configure API-auth component
            $request = new \yii\web\Request(['url' => parse_url(Yii::$app->request->getUrl(), PHP_URL_PATH)]);
            if (preg_match('/^\/api\/?+/is', $request->url)) {
                $module->controllerNamespace = 'wdmg\api\controllers\api';
                $app->setComponents([
                    'user' => [
                        'class' => '\yii\web\User',
                        'identityClass' => 'wdmg\api\models\API',
                        'enableAutoLogin' => false,
                        'enableSession' => false,
                    ], [
		                'class' => 'yii\filters\ContentNegotiator',
		                'formats' => [
			                'application/json' => yii\web\Response::FORMAT_JSON,
                            'application/xml' => yii\web\Response::FORMAT_XML,
		                ],
		                'languages' => [
			                'en',
			                'ru',
			                'uk',
		                ]
	                ]
                ]);

                $app->getUrlManager()->addRules([
                    [
                        'class' => 'yii\rest\UrlRule',
                        'controller' => [
                            'users',
                            'options',
                        ]
                    ],


                    '<module:api>/<controller:[\w-]+>/' => 'admin/<module>/<controller>',
                    '<module:api>/<controller:[\w-]+>/<action:[\w-]+>/' => 'admin/<module>/<controller>/<action>',
                    [
                        'pattern' => '<module:api>/<controller:[\w-]+>/',
                        'route' => 'admin/<module>/<controller>',
                        'suffix' => '',
                    ],[
                        'pattern' => '<module:api>/<controller:[\w-]+>/<action:\w+>/',
                        'route' => 'admin/<module>/<controller>/<action>',
                        'suffix' => '',
                    ],


                ], false);

				if (strpos(Yii::$app->request->headers['accept'], 'json') != false)
					Yii::$app->response->format = 'json';
				else if (strpos(Yii::$app->request->headers['accept'], 'xml') != false)
					Yii::$app->response->format = 'xml';

            }
        }

        // Get URL path prefix if exist
        if (isset($module->routePrefix))
            $prefix = $module->routePrefix . '/';
        else
            $prefix = '';

        // Add module URL rules
        $app->getUrlManager()->addRules(
            [
                $prefix . '<module:api>' => '<module>/api/index',
                $prefix . '<module:api>/<controller:[\w-]+>' => '<module>/<controller>',
                $prefix . '<module:api>/<controller:[\w-]+>/<action:[0-9a-zA-Z_\-]+>' => '<module>/<controller>/<action>',
                $prefix . '<module:api>/<controller:[\w-]+>/<action:[0-9a-zA-Z_\-]+>/<id:\d+>' => '<module>/<controller>/<action>',
                [
                    'pattern' => $prefix . '<module:api>/',
                    'route' => '<module>/api/index',
                    'suffix' => ''
                ], [
                    'pattern' => $prefix . '<module:api>/<controller:[\w-]+>/',
                    'route' => '<module>/<controller>',
                    'suffix' => ''
                ], [
                    'pattern' => $prefix . '<module:api>/<controller:[\w-]+>/<action:[0-9a-zA-Z_\-]+>/',
                    'route' => '<module>/<controller>/<action>',
                    'suffix' => ''
                ], [
                    'pattern' => $prefix . '<module:api>/<controller:[\w-]+>/<action:[0-9a-zA-Z_\-]+>/<id:\d+>/',
                    'route' => '<module>/<controller>/<action>',
                    'suffix' => ''
                ]
            ],
            true
        );

    }
}